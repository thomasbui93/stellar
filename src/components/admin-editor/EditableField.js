import React, {Component} from 'react'
import PropTypes from 'prop-types'
import {Editor, EditorState} from 'draft-js'

export class EditableField extends Component {
  static propTypes = {
    data: PropTypes.string,
    updateAction: PropTypes.func,
    isEditableField: PropTypes.bool,
    name: PropTypes.string
  }

  state = {
    openEditor: false
  }

  constructor (props) {
    super(props)
    this.state = {editorState: EditorState.createEmpty()}
    this.onChange = (editorState) => this.setState({editorState})
    this.toggleView = this.toggleView.bind(this)
  }

  toggleView () {
    if (this.state.openEditor) {
      this.setState({
        openEditor: false
      })
      const updateData = {}
      updateData[this.props.name] = this.refs.textInput.value
      this.props.updateAction(updateData)
    } else {
      this.setState({
        openEditor: true
      }, () => {
        this.refs.textInput.focus()
      })
    }
  }

  render () {
    const {isEditableField} = this.props
    return (
      <div className='editable-field'>
        { isEditableField
                    ? <div className='editable-field__container'>
                      <div style={{display: this.state.openEditor ? 'none' : 'block'}}
                        className='editable-field__present'
                        ref='textPresentation'
                        dangerouslySetInnerHTML={this.props.data}
                        onDoubleClick={this.toggleView} />
                      <Editor
                        style={{display: this.state.openEditor ? 'block' : 'none'}}
                        className='editable__editor'
                        editorState={this.state.editorState}
                        onChange={this.onChange}
                        onBlur={this.toggleView}
                        ref='textInput'
                        defaultValue={this.props.data} />
                    </div>
                    : <div className='editable-field__container'>
                      <div style={{display: this.state.openEditor ? 'none' : 'block'}}
                        className='editable-field__present'
                        ref='textPresentation'
                        onDoubleClick={this.toggleView}>{this.props.data}</div>
                      <input style={{display: this.state.openEditor ? 'block' : 'none'}}
                        name={this.props.name}
                        ref='textInput'
                        type='text'
                        onBlur={this.toggleView}
                        defaultValue={this.props.data} />
                    </div>
                }
      </div>
    )
  }
}
