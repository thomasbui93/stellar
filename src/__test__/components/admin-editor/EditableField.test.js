import React from 'react'
import renderer from 'react-test-renderer'
import {EditableField} from '../../../components/admin-editor/EditableField'
import {mount} from 'enzyme'
import {expect as expectChai} from 'chai'
import {spy} from 'sinon'

describe('EditableField', () => {
  const updateAction = () => {
    console.log('action update')
  }
  const key = 'test key'
  const data = 'test data'
  const isEditableField = false
  const name = 'test name'
  it('render correctly', () => {
    const component = renderer.create(
      <EditableField
        updateAction={updateAction}
        key={key}
        data={data}
        isEditableField={isEditableField}
        name={name} />
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should show input field if double click to .editable-field__present', () => {
    const component = mount(
      <EditableField
        updateAction={updateAction}
        key={key}
        data={data}
        isEditableField={isEditableField}
        name={name} />)
    component.find('.editable-field__present').simulate('doubleclick')
    expectChai(component.ref('textPresentation').style).have.property('display').equal('none')
    expectChai(component.ref('textInput').style).have.property('display').equal('block')
  })

  it('should hide input field and call update action if the input is blur', () => {
    const updateActionSpy = spy()
    const component = mount(
      <EditableField
        updateAction={updateActionSpy}
        key={key}
        data={data}
        isEditableField={isEditableField}
        name={name} />)
    component.find('.editable-field__present').simulate('doubleclick')
    component.find('input').simulate('blur')

    expectChai(component.ref('textPresentation').style).have.property('display').equal('block')
    expectChai(component.ref('textInput').style).have.property('display').equal('none')
    expectChai(updateActionSpy.called).to.equal(true)
    let expectedUpdatedData = {}
    expectedUpdatedData[name] = component.ref('textInput').value
    expectChai(updateActionSpy.calledWith(expectedUpdatedData))
  })
})
