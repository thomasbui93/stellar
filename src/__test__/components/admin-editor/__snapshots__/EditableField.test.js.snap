// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditableField render correctly 1`] = `
<div
  className="editable-field"
>
  <div
    className="editable-field__container"
  >
    <div
      className="editable-field__present"
      onDoubleClick={[Function]}
      style={
        Object {
          "display": "block",
        }
      }
    >
      test data
    </div>
    <input
      defaultValue="test data"
      name="test name"
      onBlur={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="text"
    />
  </div>
</div>
`;
